[Week7] Implement Authorization

onys-programmer opened the pull request.

onys-programmer committed:
Apply Spring Security in build.gradle

onys-programmer committed:
Create SecurityJavaConfig
Plus make csrf be disabled

onys-programmer committed:
Create JwtAuthenticationFilter
It is a filter before SpringSecurity

onys-programmer committed:
Make a filter between client and server to do authentication
Override doFilterInternal method in JwtAuthenticationFilter

onys-programmer committed:
Implement a filter in JwtAuthenticationFilter
Plus Apply it in SecurityJavaConfig

hannut91 commented:
강의를 잘 따라와 주고 계신 것 같습니다 ㅎㅎ 이번주도 화이팅입니다!

onys-programmer committed:
Apply filterWithPathAndMethod in JwtAuthenticationFilter
Plus Apply exception with no authorization

onys-programmer committed:
Create AuthenticationErrorFilter
Apply it and delete useless files

hannut91 commented:
강의 내용을 모두 학습하여 적용해보고, 테스트를 통해서 아직 구현되지 않은 곳이 확인하고, 그 다음에 개선하는 순서로 진행하면 좋을 것 같습니다

onys-programmer committed:
Try to verify authentication when product is created

onys-programmer committed:
Create UserAuthentication
Plus give it userId as a parameter

onys-programmer committed:
Make getPrincipal() method return userId
Plus add toString method in UserAuthentication

onys-programmer committed:
Set authentication when authorization is not null
In JwtAuthenticationFilter

onys-programmer committed:
Apply preAuthorize annotation at create() method in ProductController

onys-programmer committed:
Add an exceptionhandling: Unauthorized
In SecurityJavaConfig

onys-programmer committed:
Implement isAuthenticated() method in UserAuthentication
It returns true

onys-programmer committed:
Show that some methods require authentication
Apply preAuthorize annotation at each method that requires Authentication

onys-programmer committed:
Add authorization role USER in UserAuthentication
Plus Apply it in ProductController

onys-programmer committed:
Prevent unintended storage of requests
Add sessionCreationPolicy in SecurityJavaConfig
SessionCreationPolicy.STATELESS

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -37,6 +39,8 @@ UserResultData update(
     }
 
     @DeleteMapping("{id}")
+    // 누가 이걸 하는 거지? => authentication
+    // 이건 로그인해야 해! => authorization
```
`authentication` => 누가 하고 있는 건지
`authorization` => 그래서 그 사람이 할 수 있는 건지
이렇게 보시면 좋을 것 같습니다

hannut91 reviewed (commented):
[c0f7dfc](https://github.com/CodeSoom/spring-week7-assignment-1/pull/12/commits/c0f7dfc6d6fc74dfde717ad1973ac4552c0b361f)

커밋의 제목에는 진짜 의도가 무엇인지 드러나는게 좋습니다.

유저 인증을 할 때 `USER` role을 추가하라고 커밋을 작성해주셨는데, 진짜 의도는 `상품을 등록하거나 수정할 때, USER권한이 있는지 확인`하는 것이 진짜 의도 일 것 같아요.

hannut91 commented:
[cf8005b](https://github.com/CodeSoom/spring-week7-assignment-1/pull/12/commits/cf8005b645fde27e708a3bb6c4c882490426e5c8)
> Try to verify authentication when product is created

위커밋도 상품이 만들어질 때 인증을 `검사`하라는 것이 진짜 의도가 아니라, 사실은 인증된 사람만이 상품을 등록할 수 있게 하는 것이죠

onys-programmer committed:
Make changewith() method can not change password
In the User.java

onys-programmer committed:
Set default value at email, name, password in User.java
Apply it to the test for that

onys-programmer committed:
Add a test for encryption with changePassword method in UserTest

onys-programmer committed:
Implement encryption with password ar User.Java
Apply BCryptPasswordEncoder

onys-programmer commented:
> [c0f7dfc](https://github.com/CodeSoom/spring-week7-assignment-1/pull/12/commits/c0f7dfc6d6fc74dfde717ad1973ac4552c0b361f)
> 
> 커밋의 제목에는 진짜 의도가 무엇인지 드러나는게 좋습니다.
> 
> 유저 인증을 할 때 `USER` role을 추가하라고 커밋을 작성해주셨는데, 진짜 의도는 `상품을 등록하거나 수정할 때, USER권한이 있는지 확인`하는 것이 진짜 의도 일 것 같아요.

그렇군요, 사실 커밋 메시지를 정하려면 내가 지금 무슨 일을 하는 건지 생각을 많이 해야하는데 
이번주는 도무지 쉽게 이해가 되지 않아 일단 "손으로 코딩해보고 다시 한 번 살펴보자" 라는 마인드로 과제를 진행했습니다.
그 과정에서 진짜 의도를 파악하지 못한 커밋 메시지, 혹은 숲을 보지 못하는 커밋 메시지가 생겨난 것 같네요..

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -15,7 +15,7 @@ void changeWith() {
 
         assertThat(user.getName()).isEqualTo("TEST");
-        assertThat(user.getPassword()).isEqualTo("TEST");
+        assertThat(user.getPassword()).isEqualTo(null);
```
여기가 null이 되어야하는게 이해가 되지 않네요 ㅎㅎ 사용자의 정보를 수정했는데 왜 비밀번호가 null이 될까요?

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -15,7 +15,17 @@ void changeWith() {
 
         assertThat(user.getName()).isEqualTo("TEST");
-        assertThat(user.getPassword()).isEqualTo("TEST");
+        assertThat(user.getPassword()).isEqualTo("");
```
여기서 빈 문자열이 되어야하는게 이해가 되지 않네요 ㅎㅎ 사용자의 정보를 수정했는데 비밀번호는 변하지 않는걸 테스트하고 싶으셨군요 ㅎㅎ 그렇다면 이렇게 하는 것은 어떨까요?

```java
const originPassword = user.getPassword();

user.changeWith(User.builder()
                .name("TEST")
                .password("TEST")
                .build());

assertThat(user.getPassword()).isEqualTo(originPassword);
```

onys-programmer committed:
Set tokens as fields in UserControllerTest
Plus correct a parameter of claim() at encode method in JwtUtil

onys-programmer committed:
Apply MY_TOKEN at methods in UserControllerTest

onys-programmer committed:
Add a test for updateUser without accessToken
In UserControllerTest

onys-programmer committed:
Make update method can be excuted with Authentication
Apply PreAuthorize annotation at update() method in UserController

onys-programmer committed:
Add a test for forbidding users to update other's information
Add a test for updateUserWithOthersAccessToken in UserControllerTest

hannut91 commented:
이번 과제를 시작할 때 기대했던 것은 무엇이었나요? 기대했던 것을 모두 얻으셨나요?
